### PR TITLE
Transformer RatioTapChanger reactive power control

### PR DESCRIPTION
**Pull request by p-arvy**
**Please check if the PR fulfills these requirements**

- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Does this PR already have an issue describing the problem?**

#896 

**What kind of change does this PR introduce?**

Feature

**What is the current behavior?**

Transformer ratio tap changer can only regulate voltage

**What is the new behavior (if this is a feature change)?**
Transformer ratio tap changer can regulate voltage or reactive power.
Added an outerloop to control reactive power on a branch terminal with ratio tap changers.
The outerloop is implemented, as well as TransformerReactivePowerControl on branches.

**Does this PR introduce a breaking change or deprecate an API?**

No

**Other information**:

See also https://github.com/powsybl/powsybl-core/pull/2758

### CODE REVIEW

File: src/main/java/com/powsybl/openloadflow/network/PiModelArray.java [hunk 1]
```diff
@@ -316,6 +316,39 @@ public boolean shiftOneTapPositionToChangeA1(Direction direction) {
         return false;
     }
 
+    public boolean shiftOneTapPositionToChangeR1(Direction direction) {
```
Comment (annetill) [inline src/main/java/com/powsybl/openloadflow/network/PiModelArray.java:319]: Why do you need a new method?

Comment (annetill) [reply src/main/java/com/powsybl/openloadflow/network/PiModelArray.java:319]: See incremental voltage outer loop for transformer.

File: src/main/java/com/powsybl/openloadflow/network/LfBranch.java [hunk 2]
```diff
@@ -163,6 +177,10 @@ default List<LfLimit> getLimits2(LimitType type) {
 
     boolean isSpanningTreeEdge(LoadFlowModel loadFlowModel);
 
+    Evaluable getR1();
```
[the remark this answers is not in the document]
Comment (annetill) [reply src/main/java/com/powsybl/openloadflow/network/LfBranch.java:180]: How do we do for incremental voltage outer loop for transformer ?

File: src/main/java/com/powsybl/openloadflow/OpenLoadFlowParameters.java [hunk 3]
```diff
@@ -395,6 +398,8 @@ public enum ReactiveRangeCheckMode {
 
     private int incrementalTransformerVoltageControlOuterLoopMaxTapShift = IncrementalTransformerVoltageControlOuterLoop.DEFAULT_MAX_TAP_SHIFT;
 
+    private int incrementalTransformerReactivePowerControlOuterLoopMaxTapShift = IncrementalTransformerReactivePowerControlOuterLoop.DEFAULT_MAX_TAP_SHIFT;
```
[the remark this answers is not in the document]
Comment (p-arvy) [reply src/main/java/com/powsybl/openloadflow/OpenLoadFlowParameters.java:401]: Yes I agree. To do so, and to keep the code clean, shouldn't we refactor IncrementalTransformerOuterLoop classes ? We could add an intermediate mother class (between AcOuterLoop class AbstractTransformerReactivePowerControlOuterLoop/AbstractTransformerVoltageControlOuterLoop) to store common attributs/methods. What do you think @annetill @geofjamg ?

File: src/main/java/com/powsybl/openloadflow/ac/outerloop/AbstractTransformerReactivePowerControlOuterLoop.java [hunk 4]
```diff
@@ -0,0 +1,29 @@
+/**
+ * Copyright (c) 2023, RTE (http://www.rte-france.com)
```
Comment (annetill) [inline src/main/java/com/powsybl/openloadflow/ac/outerloop/AbstractTransformerReactivePowerControlOuterLoop.java:2]: fix copyright.

File: src/main/java/com/powsybl/openloadflow/OpenLoadFlowProvider.java [hunk 5]
```diff
@@ -274,6 +276,11 @@ public Optional<Extension<LoadFlowParameters>> loadSpecificParameters(Map<String
         return Optional.of(OpenLoadFlowParameters.load(properties));
     }
 
+    @Override
```
Comment (annetill) [inline src/main/java/com/powsybl/openloadflow/OpenLoadFlowProvider.java:279]: To remove.

File: .github/workflows/maven.yml [hunk 6]
```diff
@@ -25,6 +25,15 @@ jobs:
         with:
           java-version: 17
 
+      - name: Checkout powsybl-core sources
```
Comment (annetill) [inline .github/workflows/maven.yml:28]: To be removed.

File: pom.xml [hunk 7]
```diff
@@ -48,7 +48,7 @@
         <slf4jtoys.version>1.6.3</slf4jtoys.version>
         <asciitable.version>0.3.2</asciitable.version>
 
-        <powsybl-core.version>6.1.0-alpha-1</powsybl-core.version>
+        <powsybl-core.version>6.1.0-SNAPSHOT</powsybl-core.version>
```
Comment (annetill) [inline pom.xml:51]: To be changed.

File: src/main/java/com/powsybl/openloadflow/OpenLoadFlowParameters.java [hunk 8]
```diff
@@ -181,6 +182,8 @@ public enum SlackDistributionFailureBehavior {
 
     public static final String INCREMENTAL_TRANSFORMER_VOLTAGE_CONTROL_OUTER_LOOP_MAX_TAP_SHIFT_PARAM_NAME = "incrementalTransformerVoltageControlOuterLoopMaxTapShift";
 
+    public static final String INCREMENTAL_TRANSFORMER_REACTIVE_POWER_CONTROL_OUTER_LOOP_MAX_TAP_SHIFT_PARAM_NAME = "incrementalTransformerReactivePowerControlOuterLoopMaxTapShift";
```
Comment (annetill) [inline src/main/java/com/powsybl/openloadflow/OpenLoadFlowParameters.java:185]: @jeandemanged can we shared the parameter with voltage control max shift?

File: src/main/java/com/powsybl/openloadflow/network/TransformerReactivePowerControl.java [hunk 9]
```diff
@@ -0,0 +1,35 @@
+/**
+ * Copyright (c) 2023, RTE (http://www.rte-france.com)
```
Comment (annetill) [inline src/main/java/com/powsybl/openloadflow/network/TransformerReactivePowerControl.java:2]: Fix copyright here.

File: src/main/java/com/powsybl/openloadflow/network/impl/LfNetworkLoaderImpl.java [hunk 10]
```diff
@@ -637,6 +654,55 @@ private static void createTransformerVoltageControl(LfNetwork lfNetwork, RatioTa
...
+        }
+        // Check on controller branch
+        LfBranch controllerBranch = lfNetwork.getBranchById(controllerBranchId);
+        if (!controllerBranch.isConnectedAtBothSides()) {
+            LOGGER.trace("Reactive power controller branch '{}' is open: transformer reactive power control discarded", controllerBranchId);
+            report.transformerReactivePowerControlDiscardedBecauseControllerBranchIsOpen++;
+            return;
+        }
+
+        if (rtc.getRegulationTerminal().getBusView().getBus() == null) {
```
Comment (annetill) [inline src/main/java/com/powsybl/openloadflow/network/impl/LfNetworkLoaderImpl.java:670]: To remove.

File: src/main/java/com/powsybl/openloadflow/network/impl/LfNetworkLoaderImpl.java [hunk 11]
```diff
@@ -724,6 +790,10 @@ private LfNetwork create(int numCC, int numSC, Network network, List<Bus> buses,
                 // Discrete voltage controls need to be created after voltage controls (to test if both generator and transformer voltage control are on)
                 createTransformersVoltageControls(lfNetwork, parameters, loadingContext, report);
             }
+            // TODO : add check if both generator and transformer reactive power control are on
```
Comment (annetill) [inline src/main/java/com/powsybl/openloadflow/network/impl/LfNetworkLoaderImpl.java:793]: Remove TODO here.

File: src/main/java/com/powsybl/openloadflow/OpenLoadFlowParameters.java (hunk 8 above)
Comment (jeandemanged) [reply src/main/java/com/powsybl/openloadflow/OpenLoadFlowParameters.java:185]: I guess we can yes

If we share, we should rename everywhere
INCREMENTAL_TRANSFORMER_VOLTAGE_CONTROL_OUTER_LOOP_MAX_TAP_SHIFT etc
to
INCREMENTAL_TRANSFORMER_TAP_CONTROL_OUTER_LOOP_MAX_TAP_SHIFT 

reviewing code I see that for phase shifter there is no max ... we should probably do something about it

Comment (jeandemanged) [reply src/main/java/com/powsybl/openloadflow/OpenLoadFlowParameters.java:185]: I'd say 
INCREMENTAL_TRANSFORMER_RATIO_TAP_CONTROL_OUTER_LOOP_MAX_TAP_SHIFT -> for both voltage and reactive power control by RTC

later on (**not in this PR**) we would introduce 

INCREMENTAL_TRANSFORMER_PHASE_TAP_CONTROL_OUTER_LOOP_MAX_TAP_SHIFT

File: pom.xml [hunk 12]
```diff
@@ -48,7 +48,7 @@
         <slf4jtoys.version>1.6.3</slf4jtoys.version>
         <asciitable.version>0.3.2</asciitable.version>
```
Comment (Hadrien-Godard) [inline pom.xml:50]: Do not forget to remove this one

**Review by Hadrien-Godard — APPROVED** [review]

File: src/main/java/com/powsybl/openloadflow/OpenLoadFlowParameters.java [hunk 13]
```diff
@@ -239,7 +244,8 @@ private static <E extends Enum<E>> List<Object> getEnumPossibleValues(Class<E> e
         new Parameter(PLAUSIBLE_ACTIVE_POWER_LIMIT_PARAM_NAME, ParameterType.DOUBLE, "Plausible active power limit", LfNetworkParameters.PLAUSIBLE_ACTIVE_POWER_LIMIT_DEFAULT_VALUE),
         new Parameter(SLACK_BUS_P_MAX_MISMATCH_PARAM_NAME, ParameterType.DOUBLE, "Slack bus max active power mismatch", SLACK_BUS_P_MAX_MISMATCH_DEFAULT_VALUE),
         new Parameter(VOLTAGE_PER_REACTIVE_POWER_CONTROL_PARAM_NAME, ParameterType.BOOLEAN, "Voltage per reactive power slope", VOLTAGE_PER_REACTIVE_POWER_CONTROL_DEFAULT_VALUE),
-        new Parameter(REACTIVE_POWER_REMOTE_CONTROL_PARAM_NAME, ParameterType.BOOLEAN, "SVC remote reactive power control", REACTIVE_POWER_REMOTE_CONTROL_DEFAULT_VALUE),
+        new Parameter(GENERATOR_REACTIVE_POWER_REMOTE_CONTROL_PARAM_NAME, ParameterType.BOOLEAN, "SVC remote reactive power control", GENERATOR_REACTIVE_POWER_REMOTE_CONTROL_DEFAULT_VALUE),
```
Comment (jeandemanged) [inline src/main/java/com/powsybl/openloadflow/OpenLoadFlowParameters.java:247]: incorrect parameter description, to be clarified if it applies to gen only, or gen+svc, or gen+svc+vsc, ...

File: src/main/java/com/powsybl/openloadflow/OpenLoadFlowParameters.java [hunk 14]
```diff
@@ -839,14 +856,15 @@ public OpenLoadFlowParameters setDebugDir(String debugDir) {
...
 
-    public OpenLoadFlowParameters setIncrementalTransformerVoltageControlOuterLoopMaxTapShift(int incrementalTransformerVoltageControlOuterLoopMaxTapShift) {
-        this.incrementalTransformerVoltageControlOuterLoopMaxTapShift = checkParameterValue(incrementalTransformerVoltageControlOuterLoopMaxTapShift,
-                incrementalTransformerVoltageControlOuterLoopMaxTapShift >= 1,
-                INCREMENTAL_TRANSFORMER_VOLTAGE_CONTROL_OUTER_LOOP_MAX_TAP_SHIFT_PARAM_NAME);
+    public OpenLoadFlowParameters setIncrementalTransformerRatioTapControlOuterLoopMaxTapShift(int incrementalTransformerRatioTapControlOuterLoopMaxTapShift) {
+        if (incrementalTransformerRatioTapControlOuterLoopMaxTapShift < 1) {
+            throw new IllegalArgumentException("Invalid value for parameter incrementalTransformerRatioTapControlOuterLoopMaxTapShift: " + incrementalTransformerRatioTapControlOuterLoopMaxTapShift);
+        }
+        this.incrementalTransformerRatioTapControlOuterLoopMaxTapShift = incrementalTransformerRatioTapControlOuterLoopMaxTapShift;
```
Comment (jeandemanged) [inline src/main/java/com/powsybl/openloadflow/OpenLoadFlowParameters.java]: Adjust check to be like in #939

File: src/main/java/com/powsybl/openloadflow/OpenLoadFlowParameters.java [hunk 15]
```diff
@@ -1420,7 +1442,8 @@ static LfNetworkParameters getNetworkParameters(LoadFlowParameters parameters, O
                 .setPhaseControl(parameters.isPhaseShifterRegulationOn())
                 .setTransformerVoltageControl(parameters.isTransformerVoltageControlOn())
                 .setVoltagePerReactivePowerControl(parametersExt.isVoltagePerReactivePowerControl())
-                .setReactivePowerRemoteControl(parametersExt.hasReactivePowerRemoteControl())
+                .setGeneratorReactivePowerRemoteControl(parametersExt.hasGeneratorReactivePowerRemoteControl())
+                .setTransformerReactivePowerControl(parametersExt.isTransformerReactivePowerControl())
```
Comment (jeandemanged) [inline src/main/java/com/powsybl/openloadflow/OpenLoadFlowParameters.java:1445]: it was like this before already, but I don't know why sometimes we use `has` and sometimes `is`

File: src/main/java/com/powsybl/openloadflow/network/LfNetwork.java [hunk 16]
```diff
@@ -550,7 +550,7 @@ public void fix(boolean minImpedance, double lowImpedanceThreshold) {
             // zero impedance phase shifter controller or controlled branch is not supported
             branches.stream()
                     .filter(b -> b.isPhaseController() || b.isPhaseControlled()
-                            || b.getGeneratorReactivePowerControl().isPresent())
+                            || b.getGeneratorReactivePowerControl().isPresent() || b.isTransformerReactivePowerControlled())
```
Comment (jeandemanged) [inline src/main/java/com/powsybl/openloadflow/network/LfNetwork.java:553]: maybe also `|| isTransformerReactivePowerController()` ?

**Review by jeandemanged — CHANGES_REQUESTED** [review]

File: src/main/java/com/powsybl/openloadflow/network/LfNetwork.java (hunk 16 above)
Comment (annetill) [reply src/main/java/com/powsybl/openloadflow/network/LfNetwork.java:553]: Yes

File: src/main/java/com/powsybl/openloadflow/OpenLoadFlowParameters.java (hunk 15 above)
Comment (jeandemanged) [reply src/main/java/com/powsybl/openloadflow/OpenLoadFlowParameters.java:1445]: unsure of what to do, if to be done would be done in another PR, resolving

**Review by jeandemanged — APPROVED** [review]
nice! thanks!

**Pull request merged**